driver-core/src/test/functional/com/mongodb/client/model/expressions/DocumentExpressionsFunctionalTest.java

```diff
@@ -0,0 +1,238 @@
+        assertExpression(2007, ofDoc("{a: {'$date': '2007-12-03T10:15:30.005Z'}}")
+                .getDate("a", Instant.EPOCH).year(of("UTC")));
+        // no convenience for arrays
+        assertExpression(Document.parse("{b: 2}"), ofDoc("{a: {b: 2}}")
+                .getDocument("a", Document.parse("{z: 99}")));
+
+        // normal
+        assertExpression(true, ofDoc("{a: true}").getBoolean("a", of(false)));
+        assertExpression(1.0, ofDoc("{a: 1.0}").getNumber("a", of(99)));
+        assertExpression("A", ofDoc("{a: 'A'}").getString("a", of("Z")));
+        assertExpression(2007, ofDoc("{a: {'$date': '2007-12-03T10:15:30.005Z'}}")
+                .getDate("a", of(Instant.EPOCH)).year(of("UTC")));
+        assertExpression(Arrays.asList(3, 2), ofDoc("{a: [3, 2]}").getArray("a", ofIntegerArray(99, 88)));
+        assertExpression(Document.parse("{b: 2}"), ofDoc("{a: {b: 2}}")
+                .getDocument("a", of(Document.parse("{z: 99}"))));
+
+        // right branch (missing field)
+        assertExpression(false, ofDoc("{}").getBoolean("a", false));
+        assertExpression(99, ofDoc("{}").getNumber("a", 99));
+        assertExpression(99L, ofDoc("{}").getNumber("a", 99L));
+        assertExpression(99.0, ofDoc("{}").getNumber("a", 99.0));
+        assertExpression(Decimal128.parse("99"), ofDoc("{}").getNumber("a", Decimal128.parse("99")));
+        assertExpression("Z", ofDoc("{}").getString("a", "Z"));
+        assertExpression(1970, ofDoc("{}")
+                .getDate("a", Instant.EPOCH).year(of("UTC")));
+        assertExpression(Arrays.asList(99, 88), ofDoc("{}").getArray("a", ofIntegerArray(99, 88)));
+        assertExpression(Document.parse("{z: 99}"), ofDoc("{}")
+                .getDocument("a", Document.parse("{z: 99}")));
+    }
+
+    @Test
+    public void getFieldMissingTest() {
+        // missing fields
+        assertExpression(
+                BsonDocument.parse("{'a': 1}"),
+                a1.setField("z", a1.getBoolean("missing")));
+        assertExpression(
+                BsonDocument.parse("{'a': 1}"),
+                a1.setField("z", a1.getDocument("missing").getDocument("also_missing")));
+        assertExpression(
+                BsonDocument.parse("{'a': 1, 'z': ''}"),
+                a1.setField("z", a1.getString("missing").toLower()));
+        /*
+        The behaviour of missing fields appears to be as follows, and equivalent to $$REMOVE:
+        propagates -- getField, cond branches...
+        false -- not, or, cond check...
+        0     -- sum...
+        ""    -- toLower...
+        null  -- multiply, add, subtract, year, filter, reduce, map, result within map...
+        */
+    }
+
+    @Test
+    public void setFieldTest() {
+        // https://www.mongodb.com/docs/manual/reference/operator/aggregation/setField/
+        // "setField" is "withField": nothing is modified, and a new copy is produced.
+        // Placing a field based on a literal:
```

Comment:
> "setField" is "withField": nothing is modified, and a new copy is produced

The server docs state that `$setField` _"Adds, updates, or removes a specified field in a document"_. However, in all examples the documentation uses `$setField` within the `$replaceWith` pipeline stage. I.e., at least from the examples it does not look like `$setFields` modifies documents. Is the server documentation just incorrect, or are there situations when `$setField` actually modifies a document? I am guessing there are non, but maybe you know better.

Reply:
See also https://www.mongodb.com/docs/manual/reference/operator/aggregation/unset/

When it says "adds", it means "creates a copy and adds" or "returns the same document but with that field added".

It never modifies things, for the `setField` that is on the agg pipeline itself (iirc, but now I am second-guessing).

This is, of course, confusing. Especially on the aggregation pipeline itself.